File: stock-research-agent/stock_agent/research.py
"""Module 3: DeepResearch with judge scoring + no-feedback retry.

The loop borrows the keep/discard idea from autoresearch's program.md: run a
research pass, score it with an INDEPENDENT judge, and if it doesn't clear the
bar, retry — WITHOUT feeding the judge's critique back (the design explicitly
wants independent re-sampling, not "teaching to the test"). Retry diversity
comes primarily from angle rotation + retrieval-facet rotation, with a small
temperature bump as a secondary perturbation.

The research *producer* is pluggable:
- standalone: an injected callable or a plain LLM call (``LLMResearcher``),
- hermes: a ``delegate_task`` subagent (see ``hermes_tools/stock_research_tool.py``).

Entry point: ``run_deep_research(ctx, researcher, judge_llm, config)``.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Protocol

from .config import StockConfig, DEFAULT_CONFIG
from .context import ResearchContext, ResearchSlot
from .llm import LLMClient, extract_json
from .templates import render


# Angles rotated across attempts to force coverage of different info facets.
_ANGLES = [
    "最新财报、业绩指引与关键财务数据",
    "分析师/研究机构评级变化与目标价",
    "近期新闻、公告与价格异动事件",
    "供应链、需求端与行业竞争格局",
    "市场情绪、资金流向与持仓变化",
    "监管、诉讼与其他重大风险事件",
    "宏观与板块层面的驱动因子",
    "估值水平与同业对比",
]


class Researcher(Protocol):
    """Produces a research result for one attempt.

    Returns a dict: {"digest": str, "sources": [ {title,url,date}, ... ]}.
    ``angle`` and ``temperature`` are the diversity levers for this attempt.
    """

    def __call__(
        self, *, ctx: ResearchContext, angle: str, temperature: float, attempt: int
    ) -> Dict[str, Any]:
        ...


@dataclass
class LLMResearcher:
    """Simple researcher backed by a single LLM call (no live web).

    Useful standalone/testing. In production the hermes wrapper uses a
    ``delegate_task`` subagent with real web_search/web_extract tools instead.
    """

    llm: LLMClient
    model: Optional[str] = None

    def __call__(self, *, ctx, angle, temperature, attempt):
        prompt = render(
            "research_task",
            target=ctx.query.target,
            question=ctx.query.question or "(未指定)",
            horizon=ctx.query.horizon or "(未指定)",
            angle=angle,
        )
        out = self.llm.complete(
            system="You are a rigorous equity research analyst. Output JSON only.",
            user=prompt,
            model=self.model,
            temperature=temperature,
            response_format="json",
        )
        try:
            parsed = extract_json(out)
        except ValueError:
            parsed = {}
        return {
            "digest": str((parsed or {}).get("digest", "")),
            "sources": list((parsed or {}).get("sources", []) or []),
        }


def build_researcher(
    llm: LLMClient, config: StockConfig = DEFAULT_CONFIG, *, retriever=None
) -> Researcher:
    """Pick the research producer per config.

    - ``engine == "odr"``  -> multi-agent OpenDeepResearcher (default)
    - ``engine == "legacy"`` -> single-shot LLMResearcher

    ``retriever`` (ODR only) injects how each sub-topic is investigated: default
    is a single-LLM stub; hermes passes a web_search/web_extract subagent.

    Either way the OUTER judge/retry loop in ``run_deep_research`` is identical:
    a below-threshold result re-runs the whole producer — so with ODR selected,
    a low final-report score re-runs the entire ODR flow.
    """
    if config.research.engine == "odr":
        from .odr import OpenDeepResearcher

        return OpenDeepResearcher(llm=llm, config=config, retriever=retriever)
    return LLMResearcher(llm=llm, model=config.research.research_model)


def judge_digest(
    ctx: ResearchContext,
    digest: str,
    sources: List[Dict[str, Any]],
    judge_llm: LLMClient,
    config: StockConfig = DEFAULT_CONFIG,
) -> Dict[str, Any]:
    """Independent judge call. Returns {score, reasons, worst_gap}."""
    prompt = render(
        "research_judge",
        target=ctx.query.target,
        question=ctx.query.question or "(未指定)",
        horizon=ctx.query.horizon or "(未指定)",
        digest=digest or "(空)",
        source_count=str(len(sources)),
        sources=json.dumps(sources, ensure_ascii=False)[:4000],
    )
    out = judge_llm.complete(
        system="You are a strict research-quality judge. Output JSON only.",
        user=prompt,
        model=config.research.judge_model,
        temperature=0.0,
        response_format="json",
    )
    try:
        parsed = extract_json(out)
    except ValueError:
        parsed = {}
    try:
        score = float((parsed or {}).get("score", 0.0))
    except (TypeError, ValueError):
        score = 0.0
    return {
        "score": max(0.0, min(10.0, score)),
        "reasons": str((parsed or {}).get("reasons", "")),
        "worst_gap": str((parsed or {}).get("worst_gap", "")),
    }


def _persist_attempt(ctx: ResearchContext, workdir: str, attempt: int, payload: Dict[str, Any]) -> None:
    rdir = os.path.join(ctx.run_dir(workdir), "research")
    os.makedirs(rdir, exist_ok=True)
    with open(os.path.join(rdir, f"attempt_{attempt}.json"), "w", encoding="utf-8") as f:
        json.dump(payload, f, ensure_ascii=False, indent=2)


def run_deep_research(
    ctx: ResearchContext,
    researcher: Researcher,
    judge_llm: LLMClient,
    config: StockConfig = DEFAULT_CONFIG,
    *,
    persist: bool = True,
) -> ResearchContext:
    """Run the judge/retry loop and write the accepted digest into ctx.research.

    - accept as soon as score >= threshold
    - otherwise keep the best-so-far and retry (NO feedback to the researcher)
    - stop at max_attempts; if never cleared, take best-effort with a note
    """
    rc = config.research
    best: Optional[Dict[str, Any]] = None
    history: List[Dict[str, Any]] = []

    def _trace_of() -> Dict[str, Any]:
        """Pull the ODR trace for the just-finished attempt, if the producer
        exposes one (OpenDeepResearcher stores it on ``last_result``)."""
        last = getattr(researcher, "last_result", None)
        if last is not None and hasattr(last, "to_trace_dict"):
            return last.to_trace_dict()
        return {}

    for attempt in range(1, rc.max_attempts + 1):
        angle = _ANGLES[(attempt - 1) % len(_ANGLES)]
        temperature = min(
            rc.temperature_cap,
            rc.base_temperature + (attempt - 1) * rc.temperature_step,
        )

        produced = researcher(ctx=ctx, angle=angle, temperature=temperature, attempt=attempt)
        digest = str(produced.get("digest", ""))
        sources = list(produced.get("sources", []) or [])

        verdict = judge_digest(ctx, digest, sources, judge_llm, config)
        score = verdict["score"]
        trace = _trace_of()

        # Record one history entry per attempt (each is a full ODR re-run when
        # engine=="odr"): score + judge verdict + a compact trace summary.
        history.append({
            "attempt": attempt,
            "angle": angle,
            "temperature": round(temperature, 3),
            "score": score,
            "accepted": score >= rc.threshold,
            "judge": verdict,
            "supervisor_rounds": trace.get("supervisor_rounds", 0),
            "sub_topics": [f.get("topic", "") for f in trace.get("findings", [])],
        })

        if persist:
            _persist_attempt(
                ctx, config.workdir, attempt,
                {
                    "attempt": attempt, "angle": angle, "temperature": temperature,
                    "score": score, "judge": verdict,  # reasons logged, NOT reused
                    "digest": digest, "sources": sources,
                    "odr": trace,  # full ODR trace for this attempt
                },
            )

        if best is None or score > best["score"]:
            best = {"digest": digest, "sources": sources, "score": score, "odr": trace}

        if score >= rc.threshold:
            ctx.research = ResearchSlot(
                status="accepted", score=score, attempts=attempt,
                threshold=rc.threshold,
                digest=digest, sources=sources,
                engine=rc.engine, history=history, odr=trace,
            )
            if persist:
                ctx.save(config.workdir)
            return ctx
        # else: retry WITHOUT feeding verdict["reasons"] back to the researcher.

    # Exhausted attempts — take the best-effort result, flagged.
    assert best is not None
    ctx.research = ResearchSlot(
        status="best_effort", score=best["score"], attempts=rc.max_attempts,
        threshold=rc.threshold,
        digest=best["digest"], sources=best["sources"],
        note=f"未达阈值 {rc.threshold}，{rc.max_attempts} 次重试后取最优（{best['score']:.1f}）。",
        engine=rc.engine, history=history, odr=best.get("odr", {}),
    )
    if persist:
        ctx.save(config.workdir)
    return ctx

File: stock-terminal/src/data/research.ts
import type { ResearchRun } from "@/types/analysis";

// ODR 研究过程 mock —— 按股票 id 索引，对齐后端 ctx.research（engine/status/
// score/attempts/threshold/history/odr）。接后端时用 adapter 把每个 run 的
// ResearchContext.research 映射到这里即可。
//
// 000100 TCL科技：一次通过（attempt 1 即 8.6 ≥ 8.0）。
// 002185 华天科技：首轮 6.2 未达阈值 → 重跑整个 ODR → 7.1 → 再重跑 → 8.3 通过，
//   用来演示"最终评分过低就重新走 ODR 流程"的闭环。

export const researchRuns: Record<string, ResearchRun> = {
  "000100": {
    engine: "odr",
    status: "accepted",
    score: 8.6,
    attempts: 1,
    threshold: 8.0,
    digest:
      "综合各子研究：面板价格 7 月环比再涨 5%，TCL 华星大尺寸产线满产满销；北向资金连续 3 日净买入、龙虎榜机构席位现身；多家机构上调 Q3 盈利预测，一致目标价 5.60。技术面放量突破 60 日新高、多头排列。整体趋势与逻辑共振，支持持仓者继续持有，回踩不破可加仓。",
    sources: [
      { title: "群智咨询：7月面板价格月报", url: "https://example.com/panel-price", date: "2026-07-12" },
      { title: "沪深港通北向资金流向", url: "https://example.com/northbound", date: "2026-07-15" },
      { title: "TCL科技机构调研纪要", url: "https://example.com/tcl-research", date: "2026-07-14" },
    ],
    history: [
      {
        attempt: 1,
        angle: "最新财报、业绩指引与关键财务数据",
        temperature: 0.4,
        score: 8.6,
        accepted: true,
        judge: { score: 8.6, reasons: "四角度覆盖全面、来源新且可追溯，结论与证据一致", worst_gap: "情绪面样本略少" },
        supervisor_rounds: 2,
        sub_topics: ["面板价格与产线稼动率", "资金面与龙虎榜", "机构评级与目标价", "技术面结构"],
      },
    ],
    odr: {
      brief:
        "评估 TCL科技（000100）未来 1-3 个月是否值得继续持有：围绕面板价格周期、公司产线稼动、资金面、机构预期与技术面结构展开。",
      sub_questions: [
        "面板价格与行业景气度走向",
        "TCL 华星产线稼动与盈利弹性",
        "资金面（北向/龙虎榜/机构席位）",
        "机构评级、目标价与技术面结构",
      ],
      supervisor_rounds: 2,
      findings: [
        {
          topic: "面板价格与产线稼动率",
          notes:
            "群智咨询 7 月报：大尺寸面板价格环比 +5%，连续 3 个月上行；TCL 华星 t7/t9 产线满产，稼动率维持 95%+。涨价周期确立。",
          sources: [{ title: "群智咨询：7月面板价格月报", url: "https://example.com/panel-price", date: "2026-07-12" }],
          tool_calls: 4,
          reflections: ["先核对一手价格数据", "再确认公司产线是否吃到涨价，覆盖已足够"],
        },
        {
          topic: "资金面与龙虎榜",
          notes:
            "北向资金近 3 日累计净买入约 3.1 亿元；7 月 14 日龙虎榜出现机构专用席位买入居前，游资跟风。资金面偏强。",
          sources: [{ title: "沪深港通北向资金流向", url: "https://example.com/northbound", date: "2026-07-15" }],
          tool_calls: 3,
          reflections: ["交叉验证北向与龙虎榜口径一致"],
        },
        {
          topic: "机构评级与目标价",
          notes:
            "近两周 5 家机构上调 Q3 盈利预测，一致目标价升至 5.60（现价 4.95，空间约 13%），评级以增持/买入为主。",
          sources: [{ title: "TCL科技机构调研纪要", url: "https://example.com/tcl-research", date: "2026-07-14" }],
          tool_calls: 3,
          reflections: ["记录一致预期分布，避免单一机构偏差"],
        },
        {
          topic: "技术面结构",
          notes:
            "放量突破 60 日新高，均线多头排列；量能较 5 日均量放大 1.8 倍，量价齐升。回踩 4.72 为第一支撑。",
          sources: [],
          tool_calls: 2,
          reflections: ["技术面作为佐证，不作为独立结论"],
        },
      ],
      notes: [
        {
          topic: "面板价格与产线稼动率",
          compressed: "面板价 7 月 +5%（连涨 3 月），华星产线满产、稼动 95%+，涨价周期确立。",
          sources: [{ title: "群智咨询：7月面板价格月报", url: "https://example.com/panel-price", date: "2026-07-12" }],
        },
        {
          topic: "资金面与龙虎榜",
          compressed: "北向近 3 日净买入约 3.1 亿；7/14 龙虎榜机构席位买入居前，资金面偏强。",
          sources: [{ title: "沪深港通北向资金流向", url: "https://example.com/northbound", date: "2026-07-15" }],
        },
        {
          topic: "机构评级与目标价",
          compressed: "5 家机构上调 Q3 预测，一致目标价 5.60（空间约 13%），评级偏买入。",
          sources: [{ title: "TCL科技机构调研纪要", url: "https://example.com/tcl-research", date: "2026-07-14" }],
        },
      ],
    },
  },

  "002185": {
    engine: "odr",
    status: "accepted",
    score: 8.3,
    attempts: 3,
    threshold: 8.0,
    digest:
      "综合各子研究：先进封装订单回暖，华天产能利用率环比提升，行业协会预计三季度封测稼动率回升至 85%；股价回踩年线缩量止跌，抛压减轻但增量资金不足，需放量站上 13.2 确认。分析师维持增持、等待业绩拐点。结论：可轻仓试仓，等待放量确认再加。",
    sources: [
      { title: "中国半导体行业协会封测景气报告", url: "https://example.com/atp-index", date: "2026-07-13" },
      { title: "华天科技投资者关系纪要", url: "https://example.com/htkj-ir", date: "2026-07-11" },
      { title: "先进封装产业链跟踪", url: "https://example.com/adv-packaging", date: "2026-07-14" },
    ],
    history: [
      {
        attempt: 1,
        angle: "最新财报、业绩指引与关键财务数据",
        temperature: 0.4,
        score: 6.2,
        accepted: false,
        judge: { score: 6.2, reasons: "订单回暖有据，但缺少产能利用率与资金面佐证，来源偏单一", worst_gap: "产能与资金面缺口" },
        supervisor_rounds: 2,
        sub_topics: ["封测行业景气度", "先进封装订单"],
      },
      {
        attempt: 2,
        angle: "分析师/研究机构评级变化与目标价",
        temperature: 0.5,
        score: 7.1,
        accepted: false,
        judge: { score: 7.1, reasons: "补上产能利用率，评级面清晰，但技术面/资金面仍薄", worst_gap: "技术面结构未覆盖" },
        supervisor_rounds: 3,
        sub_topics: ["封测行业景气度", "先进封装订单", "产能利用率与评级"],
      },
      {
        attempt: 3,
        angle: "近期新闻、公告与价格异动事件",
        temperature: 0.6,
        score: 8.3,
        accepted: true,
        judge: { score: 8.3, reasons: "四角度齐备，技术面与基本面互证，结论审慎且可执行", worst_gap: "业绩拐点仍需下季确认" },
        supervisor_rounds: 3,
        sub_topics: ["封测行业景气度", "先进封装订单", "产能利用率与评级", "技术面与资金面"],
      },
    ],
    odr: {
      brief:
        "评估华天科技（002185）未来 1-3 个月是否为买入时机：围绕封测行业景气、先进封装订单、公司产能利用率、机构评级与技术面结构展开。",
      sub_questions: [
        "封测行业景气度与稼动率趋势",
        "先进封装订单与公司订单能见度",
        "公司产能利用率与盈利弹性",
        "机构评级、技术面与资金面",
      ],
      supervisor_rounds: 3,
      findings: [
        {
          topic: "封测行业景气度",
          notes:
            "中国半导体行业协会：6 月封测景气度环比回升，预计三季度行业稼动率回升至 85%。周期底部抬升信号明确。",
          sources: [{ title: "中国半导体行业协会封测景气报告", url: "https://example.com/atp-index", date: "2026-07-13" }],
          tool_calls: 4,
          reflections: ["先定位行业周期位置", "确认稼动率口径为行业平均"],
        },
        {
          topic: "先进封装订单",
          notes:
            "产业链跟踪：AI/HBM 带动先进封装需求，华天先进封装产线订单排至四季度，订单能见度提升。",
          sources: [{ title: "先进封装产业链跟踪", url: "https://example.com/adv-packaging", date: "2026-07-14" }],
          tool_calls: 3,
          reflections: ["区分先进封装与传统封装贡献"],
        },
        {
          topic: "产能利用率与评级",
          notes:
            "投资者关系纪要：公司产能利用率环比提升，先进封装占比上行；分析师维持增持评级，等待业绩拐点确认。",
          sources: [{ title: "华天科技投资者关系纪要", url: "https://example.com/htkj-ir", date: "2026-07-11" }],
          tool_calls: 3,
          reflections: ["补齐上一轮缺失的产能数据"],
        },
        {
          topic: "技术面与资金面",
          notes:
            "股价回踩年线（约 12.30）缩量止跌，短均线走平；成交量较前期萎缩，抛压减轻但增量资金不足，需放量站上 13.2 才确认方向。",
          sources: [],
          tool_calls: 2,
          reflections: ["技术面作为择时佐证", "强调需放量确认，避免抢跑"],
        },
      ],
      notes: [
        {
          topic: "封测行业景气度",
          compressed: "6 月封测景气环比回升，Q3 行业稼动率有望回到 85%，周期底部抬升。",
          sources: [{ title: "中国半导体行业协会封测景气报告", url: "https://example.com/atp-index", date: "2026-07-13" }],
        },
        {
          topic: "先进封装订单",
          compressed: "AI/HBM 拉动先进封装，华天订单排至 Q4，能见度提升。",
          sources: [{ title: "先进封装产业链跟踪", url: "https://example.com/adv-packaging", date: "2026-07-14" }],
        },
        {
          topic: "产能利用率与评级",
          compressed: "产能利用率环比提升、先进封装占比上行；分析师维持增持，等待业绩拐点。",
          sources: [{ title: "华天科技投资者关系纪要", url: "https://example.com/htkj-ir", date: "2026-07-11" }],
        },
        {
          topic: "技术面与资金面",
          compressed: "回踩年线缩量止跌，需放量站上 13.2 确认；增量资金暂不足，宜轻仓试仓。",
          sources: [],
        },
      ],
    },
  },
};
